OHHTTPStubs/Sources/OHHTTPStubsResponse.m

```diff
@@ -135,23 +144,47 @@ -(instancetype)initWithFileAtPath:(NSString*)filePath
                        statusCode:(int)statusCode
                           headers:(nullable NSDictionary*)httpHeaders
 {
+    NSURL *fileURL = filePath ? [NSURL fileURLWithPath:filePath] : nil;
+    self = [self initWithFileURL:fileURL
+                      statusCode:statusCode
+                         headers:httpHeaders];
+    return self;
+}
+
+-(instancetype)initWithFileURL:(NSURL *)fileURL
+                    statusCode:(int)statusCode
+                       headers:(nullable NSDictionary *)httpHeaders {
     NSInputStream* inputStream;
-    if (filePath)
+    if (fileURL)
     {
-        inputStream = [NSInputStream inputStreamWithFileAtPath:filePath];
+        // [NSURL -isFileURL] is only available on iOS 8+
+        if (![fileURL.scheme isEqualToString:NSURLFileScheme]) {
+            @throw [NSException exceptionWithName:NSInternalInconsistencyException
+                                           reason:[NSString stringWithFormat:@"Only file URLs may be passed to %@", NSStringFromSelector(_cmd)]
+                                         userInfo:nil];
+        }
+        inputStream = [NSInputStream inputStreamWithURL:fileURL];
     }
     else
     {
-        NSLog(@"%s: nil file path. Returning empty data", __PRETTY_FUNCTION__);
+        NSLog(@"%s: nil URL. Returning empty data", __PRETTY_FUNCTION__);
         inputStream = [NSInputStream inputStreamWithData:[NSData data]];
     }
     
-    NSDictionary* attributes = [NSFileManager.defaultManager attributesOfItemAtPath:filePath error:nil];
-    unsigned long long fileSize = [[attributes valueForKey:NSFileSize] unsignedLongLongValue];
-    self = [self initWithInputStream:inputStream
-                            dataSize:fileSize
-                          statusCode:statusCode
-                             headers:httpHeaders];
+    NSError *error;
+    NSNumber *fileSize;
+    [fileURL getResourceValue:&fileSize forKey:NSURLFileSizeKey error:&error];
+    
+    if (error) {
+        NSLog(@"%s: Error getting file size for URL. Returning empty data", __PRETTY_FUNCTION__);
+        inputStream = [NSInputStream inputStreamWithData:[NSData data]];
+    } else {
+        self = [self initWithInputStream:inputStream
+                                dataSize:[fileSize unsignedLongLongValue]
+                              statusCode:statusCode
+                                 headers:httpHeaders];
+    }
+    
```

Comment:
I like this version a better; there's some duplication in returning empty data but there's much less nesting:

```objc
-(instancetype)initWithFileURL:(NSURL *)fileURL
                    statusCode:(int)statusCode
                       headers:(nullable NSDictionary *)httpHeaders {
    if (!fileURL) {
        NSLog(@"%s: nil URL. Returning empty data", __PRETTY_FUNCTION__);
        return [self initWithInputStream:[NSInputStream inputStreamWithData:[NSData data]]
                         dataSize:0
                       statusCode:statusCode
                          headers:httpHeaders];
    }
    
    // [NSURL -isFileURL] is only available on iOS 8+
    if (![fileURL.scheme isEqualToString:NSURLFileScheme]) {
        @throw [NSException exceptionWithName:NSInternalInconsistencyException
                                       reason:[NSString stringWithFormat:@"Only file URLs may be passed to %@",NSStringFromSelector(_cmd)]
                                     userInfo:nil];
    }
    
    NSNumber *fileSize;
    NSError *error;
    const BOOL success = [fileURL getResourceValue:&fileSize forKey:NSURLFileSizeKey error:&error];
    
    if (!success || !fileSize) {
        NSLog(@"%s: Couldn't get file size for URL; returning empty data instead. \
The URL was: %@. \
The operation to retrieve the file size was %@. \
The error associated with that operation was: %@",
              __PRETTY_FUNCTION__, fileURL, success ? @"successful" : @"unsuccessful", error);
        return [self initWithInputStream:[NSInputStream inputStreamWithData:[NSData data]]
                                dataSize:0
                              statusCode:statusCode
                                 headers:httpHeaders];
    }
    
    return [self initWithInputStream:[NSInputStream inputStreamWithURL:fileURL]
                            dataSize:[fileSize unsignedLongLongValue]
                          statusCode:statusCode
                             headers:httpHeaders];
}
```